File: native/apple/Sources/CoreCrypto/DevelopmentNativeCryptoEngine.swift
/// Development-only crypto engine for end-to-end UI flow wiring.
///
/// IMPORTANT:
/// This implementation is intentionally non-production and is only used to
/// unblock native feature flow integration before full cryptography parity
/// (ML-KEM + AES + Argon2id) is completed.
import CryptoKit
import Foundation

private struct DevelopmentEnvelope: Codable {
    let version: UInt8
    let privateKeyBase64Url: String
    let passwordHashHex: String
    let metadata: CryptoFileMetadata
    let plaintextBase64Url: String
}

/// Non-production crypto adapter used for development flow testing.
public struct DevelopmentNativeCryptoEngine: NativeCryptoEngine {
    public init() {}

    public func encrypt(
        plaintext: [UInt8],
        password: String,
        metadata: CryptoFileMetadata,
        encryptMetadata: Bool
    ) async throws -> EncryptionResult {
        _ = encryptMetadata

        let privateKey = generateDevelopmentKey()
        let envelope = DevelopmentEnvelope(
            version: CryptoEngine.payloadVersion,
            privateKeyBase64Url: privateKey,
            passwordHashHex: passwordHashHex(password),
            metadata: metadata,
            plaintextBase64Url: encodeBase64URL(plaintext)
        )

        let encoded = try JSONEncoder().encode(envelope)
        return EncryptionResult(
            serializedPayload: [UInt8](encoded),
            privateKeyBase64Url: privateKey
        )
    }

    public func decrypt(
        serializedPayload: [UInt8],
        password: String,
        privateKeyBase64Url: String
    ) async throws -> DecryptionResult {
        let payloadData = Data(serializedPayload)
        let envelope: DevelopmentEnvelope
        do {
            envelope = try JSONDecoder().decode(DevelopmentEnvelope.self, from: payloadData)
        } catch {
            throw NativeCryptoEngineError.invalidPayload
        }

        guard envelope.version == CryptoEngine.payloadVersion else {
            throw NativeCryptoEngineError.invalidPayload
        }
        guard envelope.privateKeyBase64Url == privateKeyBase64Url else {
            throw NativeCryptoEngineError.invalidKey
        }
        guard envelope.passwordHashHex == passwordHashHex(password) else {
            throw NativeCryptoEngineError.invalidKey
        }

        guard let plaintext = decodeBase64URL(envelope.plaintextBase64Url) else {
            throw NativeCryptoEngineError.invalidPayload
        }

        return DecryptionResult(
            plaintext: plaintext,
            metadata: envelope.metadata
        )
    }

    private func generateDevelopmentKey() -> String {
        UUID().uuidString.replacingOccurrences(of: "-", with: "")
    }

    private func passwordHashHex(_ password: String) -> String {
        let digest = SHA256.hash(data: Data(password.utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    private func encodeBase64URL(_ bytes: [UInt8]) -> String {
        Data(bytes)
            .base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private func decodeBase64URL(_ value: String) -> [UInt8]? {
        var base64 = value
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let remainder = base64.count % 4
        if remainder > 0 {
            base64 += String(repeating: "=", count: 4 - remainder)
        }
        guard let data = Data(base64Encoded: base64) else {
            return nil
        }
        return [UInt8](data)
    }
}

File: native/android/core/crypto/src/main/kotlin/com/securepastebin/core/crypto/DevelopmentNativeCryptoEngine.kt
package com.securepastebin.core.crypto

import org.json.JSONObject
import java.security.MessageDigest
import java.util.Base64
import java.util.UUID

/**
 * Development-only crypto engine for end-to-end UI flow wiring.
 *
 * IMPORTANT:
 * This implementation is intentionally non-production and is only used to
 * unblock native feature flow integration before full cryptography parity
 * (ML-KEM + AES + Argon2id) is completed.
 */
class DevelopmentNativeCryptoEngine : NativeCryptoEngine {
    override suspend fun encrypt(
        plaintext: ByteArray,
        password: String,
        metadata: CryptoFileMetadata,
        encryptMetadata: Boolean,
    ): EncryptionResult {
        val privateKey = UUID.randomUUID().toString().replace("-", "")
        val envelope = JSONObject().apply {
            put("version", CryptoEngine.PAYLOAD_VERSION.toInt())
            put("privateKeyBase64Url", privateKey)
            put("passwordHashHex", passwordHashHex(password))
            put("metadata", metadataToJson(metadata))
            put("plaintextBase64Url", encodeBase64Url(plaintext))
            put("encryptMetadata", encryptMetadata)
        }

        return EncryptionResult(
            serializedPayload = envelope.toString().toByteArray(Charsets.UTF_8),
            privateKeyBase64Url = privateKey,
        )
    }

    override suspend fun decrypt(
        serializedPayload: ByteArray,
        password: String,
        privateKeyBase64Url: String,
    ): DecryptionResult {
        val envelope = try {
            JSONObject(String(serializedPayload, Charsets.UTF_8))
        } catch (_: Exception) {
            throw IllegalArgumentException("Invalid encrypted payload.")
        }

        val version = envelope.optInt("version", -1)
        if (version != CryptoEngine.PAYLOAD_VERSION.toInt()) {
            throw IllegalArgumentException("Unsupported payload version.")
        }

        val expectedKey = envelope.optString("privateKeyBase64Url")
        if (expectedKey != privateKeyBase64Url) {
            throw IllegalArgumentException("Private key fragment does not match payload.")
        }

        val expectedPasswordHash = envelope.optString("passwordHashHex")
        if (expectedPasswordHash != passwordHashHex(password)) {
            throw IllegalArgumentException("Invalid password.")
        }

        val plaintextBase64Url = envelope.optString("plaintextBase64Url")
        val plaintext = decodeBase64Url(plaintextBase64Url)
            ?: throw IllegalArgumentException("Encrypted payload plaintext is invalid.")

        val metadataJson = envelope.optJSONObject("metadata")
            ?: throw IllegalArgumentException("Encrypted payload metadata is invalid.")

        return DecryptionResult(
            plaintext = plaintext,
            metadata = metadataFromJson(metadataJson),
        )
    }

    private fun passwordHashHex(password: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(password.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { byte -> "%02x".format(byte) }
    }

    private fun encodeBase64Url(bytes: ByteArray): String {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)
    }

    private fun decodeBase64Url(value: String): ByteArray? {
        return try {
            Base64.getUrlDecoder().decode(value)
        } catch (_: Exception) {
            null
        }
    }

    private fun metadataToJson(metadata: CryptoFileMetadata): JSONObject {
        return JSONObject().apply {
            put("name", metadata.name)
            put("size", metadata.size)
            put("mimeType", metadata.mimeType)
            put("uploadDate", metadata.uploadDate)
            if (metadata.expirationDate == null) {
                put("expirationDate", JSONObject.NULL)
            } else {
                put("expirationDate", metadata.expirationDate)
            }
            put(
                "encryptionConfig",
                JSONObject().apply {
                    put("encryptMetadata", metadata.encryptionConfig.encryptMetadata)
                    put("algorithm", metadata.encryptionConfig.algorithm)
                },
            )
        }
    }

    private fun metadataFromJson(json: JSONObject): CryptoFileMetadata {
        val encryptionConfig = json.getJSONObject("encryptionConfig")
        return CryptoFileMetadata(
            name = json.getString("name"),
            size = json.getInt("size"),
            mimeType = json.getString("mimeType"),
            uploadDate = json.getLong("uploadDate"),
            expirationDate = if (json.isNull("expirationDate")) null else json.getLong("expirationDate"),
            encryptionConfig = CryptoFileMetadata.EncryptionConfig(
                encryptMetadata = encryptionConfig.getBoolean("encryptMetadata"),
                algorithm = encryptionConfig.getString("algorithm"),
            ),
        )
    }
}
